### ISSUE TITLE
Plone/ZCA memory leak?

### ISSUE THREAD

**Issue by mauritsvanrees** [plone/Products.CMFPlone#3829]
TLDR: There seems to be a memory problem in the Zope Component Architecture.

## Background:

- A customer that we moved from Plone 5.1 to Plone 5.2 on Python 3.8 in January, experienced that the memory usage of Plone kept growing and they needed to restart. Restart was handled by systemd after a memory limit was hit. Usually a few times per week.
- The sysadmin changed the setup to start Plone with jemalloc: http://jemalloc.net/. This helped, and the memory usage stabilized. jemalloc should help when memory fragmentation was the problem.
- We switched to Plone 6.0.6 on Python 3.10 in June, and the problems are back. This is a multilingual site, with 13 languages.
- My personal site is experiencing memory growth as well. Plone 6.0, Python 3.11. My brother had to restart our shared server, or the Plone process, during my vacation because my Plone site used too much memory. Now after 17 days the memory has grown to 1.1 GB. This seems strange: the Data.fs is 30MB, and the blobstorage 59 MB. This is Plone 6.0 Classic UI, without any add-ons, and even without any "core add-ons" activated. It does support two languages, English and Dutch, but no `plone.app.multilingual` setup.

## Reproduce:

This weekend we noticed that a lot of traffic at the customer site was for one bot, and most of this went to the search pages. So I started hammering my local setup in a similar way. For every language, get the search page, and get the next 50 batched pages. Do this 100 times:

```
for j in $(seq 100); do
    for start in $(seq 50); do
        echo $j - $start
        for lang in en de es fr it pl fr-ca ar zh ja ru; do
            curl -s -o search.html "http://localhost:8080/Plone/${lang}/@@search?b_start%3Aint=${start}0&SearchableText="
        done
    done
done
```

The search url above is shortened for clarity. In reality I pasted the really long line that you get when you go to the second result page. All these languages contain at least 500 search result pages.

With this, I *did* notice that memory kept creeping up.
I radically simplified the search page to display almost nothing and basically only do the correct catalog query, and even then memory increased.

## Investigation in Plone:

I went to the Zope debug control panel on the local site that I had hammered: http://localhost:8080/Control_Panel/DebugInfo/manage_main
It showed this:
![Screenshot 2023-07-25 at 12 35 53](https://github.com/plone/Products.CMFPlone/assets/210587/add2ca44-5745-4c1c-af1d-a4dc2969e78e)

(When I look at the live site, the top two are 400,000 DateTime objects, and then 84,000 ZCatalog Benchmarks.)

Why are the counts for `NegotiateLanguage` and `DexterityPublishTraverse` so high? I focused on the first one.
It is [this class](https://github.com/plone/plone.i18n/blob/5.0.1/plone/i18n/negotiate/negotiate.py#L8). Not an _instance_ of the class, but the _actual_ class.  This is [registered as adapter](https://github.com/plone/plone.i18n/blob/2f21b4195b6beb80fe9ab1f16ba511d16443201b/plone/i18n/configure.zcml#L9-L13).  It is looked up as multi adapter in [`setLanguageBinding`](https://github.com/plone/plone.i18n/blob/2f21b4195b6beb80fe9ab1f16ba511d16443201b/plone/i18n/utility.py#L45).

## Investigation in ZCA:

Let's get some numbers. Edit `zope.interface/adapter.py` and add `import sys` at the top. Go down to the [`queryMultiAdapter`](https://github.com/zopefoundation/zope.interface/blob/6.0/src/zope/interface/adapter.py#L836-L845) method and add a few lines before returning the result: 

```
if (
    factory.__name__ in ("NegotiateLanguage", "DexterityPublishTraverse")
    or sys.getrefcount(factory) > 100
):
    print(f"Factory '{factory.__name__}' refcount={sys.getrefcount(factory)}")
```

Careful: typos may be silently ignored, depending on who calls this code.

This shows the reference count to the factory that was looked up.  See https://stackify.com/python-garbage-collection/ for an article on Python garbage collection that explains `sys.getrefcount`.

Restart Plone, and browse the site or use the curl command, and sure, increasing reference counts are printed. With my big curl command and my hacked up search page, I only see the two that I was looking at:

```
Factory 'NegotiateLanguage' refcount=13
Factory 'NegotiateLanguage' refcount=14
Factory 'NegotiateLanguage' refcount=14
Factory 'NegotiateLanguage' refcount=15
Factory 'DexterityPublishTraverse' refcount=13
Factory 'NegotiateLanguage' refcount=16
Factory 'DexterityPublishTraverse' refcount=14
Factory 'NegotiateLanguage' refcount=16
Factory 'DexterityPublishTraverse' refcount=14
Factory 'NegotiateLanguage' refcount=17
Factory 'DexterityPublishTraverse' refcount=15
Factory 'NegotiateLanguage' refcount=18
Factory 'DexterityPublishTraverse' refcount=16
Factory 'NegotiateLanguage' refcount=18
Factory 'DexterityPublishTraverse' refcount=16
Factory 'NegotiateLanguage' refcount=19
Factory 'DexterityPublishTraverse' refcount=17
Factory 'NegotiateLanguage' refcount=20
Factory 'DexterityPublishTraverse' refcount=18
Factory 'NegotiateLanguage' refcount=20
Factory 'DexterityPublishTraverse' refcount=18
```

I still had a copy of this customer site on Plone 5.2 as well, and tried the same with curl or just browsing and quickly reloading. NegotiateLanguage got a refcount of 28 and stayed there. DexterityPublishTraverse may not have existed then.  But others showed up, for example two very often used ones:
```
Factory 'PortalState' refcount=1445
Factory 'ContextState' refcount=2080
```
Sometimes counts stayed the same, or even decreased a bit. Maybe if you make the get/query multi adapter call several times in the same request, it stays the same.

## Now what?

So now the question is: is this a memory leak deep down in the Zope Component Architecture?

And how do we solve this?

Working theory:

* `queryMultiAdapter` gives you a new factory for each request.
* The old factory is never cleaned up from some component cache.

**Comment by davisagli** [comment]
@mauritsvanrees Maybe https://docs.python.org/3/library/gc.html#gc.get_referrers can be helpful to check what objects hold the references?

https://github.com/bloomberg/memray could also be helpful.

**Comment by davisagli** [comment]
It looks like instances of NegotiateLanguage store a reference to the request, and also get stored in an attribute of the request. That sounds suspicious for creating a cycle that can't be garbage collected, but it should be cleaned up when the request's clear method is called at the end of processing the request, so this might not be relevant.

**Comment by d-maurer** [comment]
`zope.interface` does important caching -- but as far as I know not for adapters.

I would continue to analyze as follows:
For `zope.testrunner` I have implemented a cycle detection (--> `gc-after-test` with verbosity level 4 or higher).
You could write a test reproducing the problem and then use this feature to determine the cycle.
Alternatively, you could take over the relevant code from `zope.testrunner` and use it in your setup. You find it in `zope.testrunner.runner` starting with
```python
        if (uses_refcounts and self.options.gc_after_test and
                self.options.verbose >= 4):
```
It essentially uses the `DiGraph` class from `zope.testrunner.digraph` (which contains an implementation for Tarjan's algorithm to detect the strongly connected graph components (the non trivial ones are the cycles)).

**Comment by d-maurer** [comment]
> Maybe only multi adapters. They keep increasing on every request.

I do not think the ZCA (more precisly `zope.interface`) responsible: multi adapters are extremely frequent (any view is a multi adapter); memory leaks related to multi adapters in general should be fatal within very short timeframes.

Potentially, the problem is limited to adapters registered in a (local) persistent adapter registry (i.e. site specific adapters). I still do not believe `zope.interface` responsible because it manages only the registrations with the adapter factories; there is no reason whatsoever to cache the adapters themselves.

**Comment by mauritsvanrees** [comment]
I don't see anything bad in the related Plone code yet, so pointing to the ZCA may be logical.  But the fact that this is not a wide spread problem for all multi adapters, indeed shows that the problem may be elsewhere.

Thanks for the hints.  So that is the `zope.testrunner` code surrounding [these lines](https://github.com/zopefoundation/zope.testrunner/blob/2c2c5ad016d1b98f2c6182dfa632e4531aaea9c7/src/zope/testrunner/runner.py#L1026-L1027).  Looks interesting.  I hope to try it out next week.

**Comment by d-maurer** [comment]
Maurits van Rees wrote at 2023-7-28 13:11 -0700:
> ...
>Thanks for the hints.  So that is the `zope.testrunner` code surrounding [these lines](https://github.com/zopefoundation/zope.testrunner/blob/2c2c5ad016d1b98f2c6182dfa632e4531aaea9c7/src/zope/testrunner/runner.py#L1026-L1027).  Looks interesting.  I hope to try it out next week.

Note that `zope.testrunner` looks only for cycles in garbarge.
It will not detect non garbage cycles.
It will not detect problems related to caching.

In addition: Python 3 is (in almost all cases) able to release cyclic
garbage (the restriction "cannot release cycles when one of its
objects has a `__del__` method" has been removed for almost all cases).
Therefore, it is not likely that cycles are still responsible
for memory leaks.

Therefore, I withdraw my previous recommendation (use cycle analysis).
Likely, `gc.get_referrers` is the better approach to find out
which structures contain the references to the leaked objects.

**Comment by d-maurer** [comment]
I agree with you that the referrer at the end of "https://github.com/plone/Products.CMFPlone/issues/3829#issuecomment-1655741632" is very suspicious and points towards the ZCA.It seems to be a cache to speed up the adapter factory lookup.

I suggest that you compare 2 to the keys in the dict. Obviously, they are different. If the corresponding tuple components are different, expand them into their list of interfaces and compare those lists. They, too, should be different; identify the components that make the lists different.

Dynamic interface definitions could lead to unlimited growth of the cache above. When I remember right, `z3c.form` might define interfaces dynamically (for instance specific customizations).

Another potential cause could be inconsistent "iro"s (= "Interface Resolution Order"). The "iro" is for interface specifications the analog to the "mro" for class definitions. It determines the order of component interfaces of an interface specification. Of course, each component interface should appear exactly once and if component interface ID is derived from interface IB, then ID should appear before IB and if component interface ID is derived from interfaces IB1 and IB2 (and potentially further interfaces) and IB1 precedes IB2 in the derivation order, then IB1 must precede IB2 in the "iro". Of course, those requirements can not always be met. `zope.interface` has an option `STRICT_IRO` which causes an exception in that case, but the "iro"s of some Zope objects violate the condition and Zope cannot use this option. As a consequence, we might observe inconsistent "iro"s.

I hope that your key analysis of the cache reveals the cause for the cache growth. more precisely why requests which should lead to equal interface specifications, apparently use different interface specifications.

**Comment by d-maurer** [comment]
> I suggest that you compare 2 to the keys in the dict.

I might be necessary to compare more than 2 keys.

I used the following script to try to reproduce and analyze the problem in my local Plone 5.2.10 instance:
```python
from sys import path; path.append(".")
from wrap import wrap
from pprint import pprint as pp
from plone.i18n.interfaces import INegotiateLanguage
from plone.i18n.negotiate.negotiate import NegotiateLanguage
from zope.component import getMultiAdapter
from gc import get_referrers as gr

def filter(l):
  """filter list *l* removing uninteresting things."""
  def cond(x):
    # remove module dicts
    if isinstance(x, dict) and "__loader__" in x: return True
    # remove non dicts
    if not isinstance(x, dict): return True
  return [x for x in l if not cond(x)]

def chk():
  global app
  app = wrap(app); R = app.REQUEST
  v = R.traverse("partner/bp3")
  os = app.partner.bp3, R
  getMultiAdapter(os, INegotiateLanguage)
  return filter(gr(NegotiateLanguage))

pp(chk())
pp(chk())
```
The used `wrap` is similar to `Testing.make_request.make_request` but ensures that a `Manager` authorization happens. This is necessary as my portal is "members only".
My portal is at `partner/bp3`.

The script was not able to reproduce the problem. But it showed that  the `INegotiateLanguage` adapter lookup was used for 2 different request interface signatures:
```pycon
[<InterfaceClass plone.app.z3cform.interfaces.IPloneFormLayer>, <InterfaceClass plone.app.discussion.interfaces.IDiscussionLayer>, <InterfaceClass bfd.bp_backend.customercontent.pdfpreview.interfaces.IPreviewLayer>, <InterfaceClass plone.app.event.interfaces.IBrowserLayer>, <InterfaceClass plone.app.theming.interfaces.IThemingLayer>, <InterfaceClass plone.app.contenttypes.interfaces.IPloneAppContenttypesLayer>, <InterfaceClass plonetheme.barceloneta.interfaces.IBarcelonetaLayer>, <InterfaceClass zope.publisher.interfaces.browser.IDefaultBrowserLayer>, <InterfaceClass zope.publisher.interfaces.browser.IBrowserRequest>, <InterfaceClass zope.annotation.interfaces.IAttributeAnnotatable>]
>>> l1
[<InterfaceClass bfd.bfd_content.browser.interfaces.IThemeSpecific>, <InterfaceClass plone.app.z3cform.interfaces.IPloneFormLayer>, <InterfaceClass plone.app.discussion.interfaces.IDiscussionLayer>, <InterfaceClass bfd.bp_backend.customercontent.pdfpreview.interfaces.IPreviewLayer>, <InterfaceClass plone.app.event.interfaces.IBrowserLayer>, <InterfaceClass plone.app.theming.interfaces.IThemingLayer>, <InterfaceClass plone.app.contenttypes.interfaces.IPloneAppContenttypesLayer>, <InterfaceClass plonetheme.barceloneta.interfaces.IBarcelonetaLayer>, <InterfaceClass plone.theme.interfaces.IDefaultPloneLayer>, <InterfaceClass zope.publisher.interfaces.browser.IDefaultBrowserLayer>, <InterfaceClass zope.publisher.interfaces.browser.IBrowserRequest>, <InterfaceClass zope.annotation.interfaces.IAttributeAnnotatable>]
>>> set(l1) - set(l0)
{<InterfaceClass bfd.bfd_content.browser.interfaces.IThemeSpecific>, <InterfaceClass plone.theme.interfaces.IDefaultPloneLayer>}
```
Something adds additional interfaces to the request after the first adapter lookup. The second lookup then creates a new cache entry for the extended request interface set.

Should something like this also happen for you, you will need to compare additional keys.

**Comment by mamico** [comment]
My 5 cents.

I checked two different key, and I saw differences only on the first tuple element: `classImplements(?)`. Loking into `zope.interface` code, I guess that this is the `providedBy` of the Plone Site Root.

So in a test view I memoized the `providedBy` of the portal and another content like this:

```
_p = []
_n = []
class A(BrowserView):
    def __call__(self):
        _p.append(providedBy(api.portal.get()))
        _n.append(providedBy(api.portal.get().news))  # tested also with other kind of content

        if len(_p) == 2:
            print(f"== p {_p[0] == _p[1]}")  # == p False
            print(f"== n {_n[0] == _n[1]}")  # == n True

            print(f"__eq__ p {_p[0].__eq__(_p[1])}") # __eq__ p NotImplemented
            print(f"__eq__ n {_n[0].__eq__(_n[1])}") # __eq__ n True

            print(f"hash p {hash(_p[0])} {hash(_p[1])} {hash(_p[0])==hash(_p[1])}")  # hash p 8742003969974 8742004544951 False
            print(f"hash n {hash(_n[0])} {hash(_n[1])} {hash(_n[0])==hash(_n[1])}")  # hash n 8742001634995 8742001634995 True

        return "A"
```

The second time I call the view, I see that the `providedBy` of the portal has "changed," otherwise the `providedBy` of the other content has not.

I also saw a `NotImplemented` on the `__eq__` method in the first case.

I hope it helps.

**Comment by d-maurer** [comment]
Mauro Amico wrote at 2023-8-13 08:41 -0700:
>I checked two different key, and I saw differences only on the first tuple element: `classImplements(?)`. Loking into `zope.interface` code, I guess that this is the `providedBy` of the Plone Site Root.

Strange: the `classImplements` should indicate that the portal
object has no `alsoProvides` applied, i.e. that the provided
interfaces come only from the class definition.
I would expect that the class definition is not changed at runtime,
at least not regarding the implemented interfaces.
In this case, the `classImplements` should not change.

You could expand the `classImplements` into the list of interfaces
(applying `list`) and analyze the differences.

**Comment by mamico** [comment]
> You could expand the `classImplements` into the list of interfaces (applying `list`) and analyze the differences.

This is what I see:

```
(Pdb) _p[0] == _p[1]
False

(Pdb) len([i for i in _p[0].flattened()])
75
(Pdb) [i for i in _p[0].flattened()]
[<InterfaceClass plone.dexterity.schema.generated.Plone_5_1691832663_2_5337133_0_Plone_1_Site>, <InterfaceClass plone.app.contenttypes.behaviors.richtext.IRichText>, ...]
(Pdb) l0 = set([i for i in _p[0].flattened()])
(Pdb) l1 = set([i for i in _p[1].flattened()])
(Pdb) l0 == l1
True

(Pdb) len([i for i in _p[0].dependents.items()])
1
(Pdb) [i for i in _p[0].dependents.items()] == [i for i in _p[1].dependents.items()]
True

Pdb) len([i for i in _p[0].interfaces()])
48
(Pdb) [i for i in _p[0].interfaces()]
[<InterfaceClass plone.dexterity.schema.generated.Plone_5_1691832663_2_5337133_0_Plone_1_Site>, <InterfaceClass plone.app.contenttypes.behaviors.richtext.IRichText>, ...]
(Pdb) l0 = [i for i in _p[0].interfaces()]
(Pdb) l1 = [i for i in _p[1].interfaces()]
(Pdb) l0 == l1
True
```

I can't figure out how the equality operator is working. My suspicions are about the order in `IRO` or something that is missing (as I mentioned before, the `__eq__` method returns `NotImplemented` here).

```
(Pdb) _p[0].__eq__(_p[1])
NotImplemented

(Pdb) _n[0].__eq__(_n[1])
True
```

**Comment by d-maurer** [comment]
Mauro Amico wrote at 2023-8-14 02:06 -0700:
>> You could expand the `classImplements` into the list of interfaces (applying `list`) and analyze the differences.
> ...

I think I found a bug which can explain the cache growth:
`plone.dexterity.content` computes a value for the `__provides__`
attribute by combining the original `__provides__` with
the interfaces from registered hehaviors via a call to
`zope.interface.declarations.Implements`.
This implements behaves strangely:
```pycon
>>> Implements
<class 'zope.interface.declarations.Implements'>
>>> Implements(*dp) == Implements(*dp)
False
```
i.e. applied to identical arguments, the result of `Implements` is not equal.

This also holds for a simple case:
```pycon
>>> class I1(Interface): pass
...
>>> Implements(I1) == Implements(I1)
False
```

`plone.dexterity` cashes the computed `__provided__` in a volatile
attribute (--> `_v__providedBy__`). Therefore, not every
lookup results in adapter cache growth. However, the cache is lost
when the object is ghostified (--> `_p_deactivate`).
In that case, the next adapter lookup created a new
adapter cache entry.


I will try to find out why `Implements` behaves this way.

**Comment by d-maurer** [comment]
> I will try to find out why `Implements` behaves this way.

A comment in `zope.interface.declarations.Implements` indicates that identity based hashing and equality is wanted. Thus, the behavior is correct and `plone.dexterity` uses `Implements` wrongly: `Implements` is used for classes, i.e. for mostly static objects; there is a different set of declarations which target instances (--> `Provides`) and which cache created declarations (in a `WeakValueDict`). Almost surely, `plone.dexterity` should use `Provides` (instead of `Implements`) to avoid the cache growth.

**Comment by mamico** [comment]
I implemented @d-maurer's suggestion (my kudos) at https://github.com/plone/plone.dexterity/pull/187.

Using a benchmark test similar to what @mauritsvanrees said earlier (6 languages, 10,000 requests per language, one thread), measuring RSS with collective.stats, the difference is quite impressive

![image](https://github.com/plone/Products.CMFPlone/assets/182599/f8de5bd6-1f53-495b-95c6-7db800f3bbfc)

**Comment by mauritsvanrees** [comment]
I cannot test it myself this week, but sounds great! Thank you both.

**Comment by mauritsvanrees** [comment]
Finally managed to test this locally.  I hammered the search page with the command under the "Reproduce" section in the main issue comment, stopping after over 100,000 requests. The memory usage stays around 220 MB.

And in the debug control panel under the top 100 reference counts, the `plone.i18n` and `plone.dexterity` classes do not even show up.

This is awesome!
Thanks a lot @d-maurer and @mamico!

I have released `plone.dexterity` 3.0.3 with this fix.  This will be released in Plone 6.0.7, expected within 2-3 weeks.

**Comment by mauritsvanrees** [comment]
The fix is live on a website since a few days.  Results are looking good:

![plone-memory-dexterity](https://github.com/plone/Products.CMFPlone/assets/210587/891463dd-d562-41a0-8027-ac4742397095)

Memory for both zeoclients used to go steadily up till 2 GB, then get automatically restarted.  After the update it stays well below 1 GB.

Thanks again!

**Comment by gforcada** [comment]
Since 2012 when we did a rewrite of our website we have been seeing this memory pattern, but never had the time to dive into it.

I can say that indeed, this works like magic ✨ 💯 

THANKS! to all involved in getting this reported/investigated/fixed! 🙇🏾

**Comment by mauritsvanrees** [comment]
We may want to back port this to Plone 5.2, which is scheduled to get its last maintenance release in October.  But I do want to wait a bit and give people the chance to detect any regressions in Plone 6.

**Comment by mauritsvanrees** [comment]
> We may want to back port this to Plone 5.2, which is scheduled to get its last maintenance release in October. But I do want to wait a bit and give people the chance to detect any regressions in Plone 6.

I checked the Debug Information on a multilingual 5.2 site, but the `plone.i18n` utility did not show up in the first 100 items.  The dexterity one does have a couple thousand items, but it does not seem to go up often, and others classes have much higher numbers.  So 5.2 seems not affected.  The dexterity site root on 6.0 may have been a trigger.

Anyway, here is another screen shot of memory usage for the Plone 6 site:

![sfia-memory-2](https://github.com/plone/Products.CMFPlone/assets/210587/4f7861a6-3b01-4aa2-844a-997269a50c80)

On the left of the graph is the old situation.  Then the new code is used.  Then in the middle two manual restarts that were needed for unrelated code.  And in all it looks much healthier, at half the memory usage.  As the admin says: "Thank you Plone team!"

**Comment by gforcada** [comment]
I have some different experience, we are still on 5.2 and 12 years ago we started with 4.3(?) I guess, but we always have been using dexterity, and not a multilingual site, but nonetheless our memory consumption was spiky until I backported this fix.

So 5.2 is indeed affected, maybe only if you are heavily using Dexterity though? 🤔

**Comment by d-maurer** [comment]
Gil Forcada Codinachs wrote at 2023-9-25 01:13 -0700:
>I have some different experience, we are still on 5.2 and 12 years ago we started with 4.3(?) I guess, but we always have been using dexterity, and not a multilingual site, but nonetheless our memory consumption was spiky until I backported this fix.
>
>So 5.2 is indeed affected, maybe only if you are heavily using Dexterity though? ?

The observed memory leak is usually hidden by a cache in
an `_v_` attribute.
It will be observed only if this cache is frequently invalidated,
either because the hosting object is often modified or for other
reasons often removed from the ZODB cache (e.g. ghostified
due to cache contention).
